extension/observer/dockerobserver/extension.go

```diff
@@ -84,29 +86,89 @@ func (d *dockerObserver) Shutdown(ctx context.Context) error {
 
 // ListAndWatch provides initial state sync as well as change notification.
 // Emits initial list of endpoints loaded upon extension Start. It then goes into
-// a loop to sync the container cache periodically and change endpoints.
-// TODO: Watch docker events to notify listener of changed endpoints as
-// events stream
+// a loop to sync the container cache periodically, watch for docker events,
+// and notify the observer listener of endpoint changes.
 func (d *dockerObserver) ListAndWatch(listener observer.Notify) {
 	d.emitContainerEndpoints(listener)
-
 	go func() {
-		ticker := time.NewTicker(d.config.CacheSyncInterval)
-		defer ticker.Stop()
+		cacheRefreshTicker := time.NewTicker(d.config.CacheSyncInterval)
+		defer cacheRefreshTicker.Stop()
+		// timestamp to begin looking back for events
+		lastTime := time.Now()
+		var eventErr error
+
+		// loop for re-connecting to events channel if a client error occurs
+		for {
+			// watchContainerEvents will only return when an error occurs, or the context is cancelled
+			lastTime, eventErr = d.watchContainerEvents(listener, cacheRefreshTicker, lastTime)
+
+			// We are only interested in re-connecting when the context hasn't been canceled
+			// since requests made with a closed context are guaranteed to fail.
+			if eventErr != nil && d.ctx.Err() == nil {
+				// Either decoding or connection error has occurred, so we should resume the event loop after
+				// waiting a moment.  In cases of extended daemon unavailability this will retry until
+				// collector teardown or background context is closed.
+				d.logger.Error("Error watching docker container events, attempting to retry", zap.Error(eventErr))
+				time.Sleep(clientReconnectTimeout)
+			} else {
+				// context was cancelled
+				return
+			}
+		}
+	}()
+}
+
+func (d *dockerObserver) watchContainerEvents(listener observer.Notify, cacheRefreshTicker *time.Ticker, lastEventTime time.Time) (time.Time, error) {
+	// build filter list for container events
+	filters := dfilters.NewArgs([]dfilters.KeyValuePair{
+		{Key: "type", Value: "container"},
+		{Key: "event", Value: "destroy"},
+		{Key: "event", Value: "die"},
+		{Key: "event", Value: "pause"},
+		{Key: "event", Value: "rename"},
+		{Key: "event", Value: "stop"},
+		{Key: "event", Value: "start"},
+		{Key: "event", Value: "unpause"},
+		{Key: "event", Value: "update"},
+	}...)
+
+	for {
```

Comment:
Given the change to return errors and timestamps I don't think this loop is necessary?

Reply:
You are correct, removed.